### ISSUE TITLE
Support for constants using IDE code completion (pycharm, eclipse, sublimecodeintel)

### ISSUE THREAD

**Issue by mtimm** [datacenter/cobra#62]

**Comment by JonasKs** [comment]
Any status on this @rchamarthy / @sganapathy85 / @sapshah-cisco ?

**Comment by JonasKs** [comment]
Or suggestions for implementation?

**Comment by giovannimeo** [comment]
@JonasKs let me see if i understand the requirement correctly, let say you have:
from cobra.model.fv import Tenant
mo = Tenant(name="test")
then while typing 

mo. 

you want to see what can complete.
Is that correct statement?
I think the problem lays in the way acimodel library is structured, here is a layout of the acimodel wheel file:
![image](https://user-images.githubusercontent.com/1607488/145977394-b0f0e65a-93b1-42ea-8b75-a5d41a4df831.png)
when you import cobra.model.fv you are actually importing a file that will then automatically load the package fv. The actual implementation of the package is in cobra.modelimpl.fv probably that is what is messing up with the completion.
The reason for this intermediate step is to avoid to load in memory all the classes even the ones that are not needed.
The trick is documented into:
https://mail.python.org/pipermail/python-ideas/2012-May/014969.html

Bottom line with this kind of redirection, i'm not sure there is much can be done, because it would mean the IDE should be ready to handle this indirection.

**Comment by JonasKs** [comment]
Thanks for getting back to me, @giovannimeo !

Yeah, pretty much. PyCharm also complains on that the module don't exist, which is very annoying.  

I've experimented a bit, and there is a few ways that _could_ improve this.

### Option 1:
Create a stub file, e.g. `fv.pyi` and add this content:

```py
class Tenant:
    annotation: str
    childAction: str
    descr: str
    dn: str
    extMngdBy: str
    lcOwn: str
    modTs: str
    monPolDn: str
    name: str
    nameAlias: str
    ownerKey: str
    ownerTag: str
    rn: str
    status: str
    uid: str
``` 
![bilde](https://user-images.githubusercontent.com/5310116/145981097-959bcaf2-231c-4a5b-8d23-086b38fc1263.png)

Not sure how this will be on memory impact, considering modelimpl is a 600mb big folder though..

### Option 2:
Import `modelimpl` for type-checking:
```py
from typing import TYPE_CHECKING

from cobra.model.fv import Tenant

if TYPE_CHECKING:
    from cobra.modelimpl.fv.tenant import Tenant

mo = Tenant(name="test")
mo.
```
will result in this:

![bilde](https://user-images.githubusercontent.com/5310116/145983279-d86bfd5d-2505-471c-9ef8-3cd20f55c038.png)


and not interfere with runtime. How ever, what I'd like to do is know which children an object can have, which properties I can set (e.g. `descr`) without having to browse the object store.

```py
mo = Tenant(uniMo, name="test")
print(mo.meta.props.names)
# prints: ['annotation', 'childAction', 'descr', 'dn', 'extMngdBy', 'lcOwn', 'modTs', 'monPolDn', 'name', 'nameAlias', 'ownerKey', 'ownerTag', 'rn', 'status', 'uid']
mo.descr = 'test descr'
``` 

As of right now, I don't really see the benefit of using `cobra` above the API directly, even though I'm certain I would if these things were better.

**Comment by giovannimeo** [comment]
@JonasKs thanks for the clarification, and to be honest with you i totally see your point, but the issue is that the ACI model if threated as code lead to a very huge repository as you correctly highlighted in your option 1 before, which was indeed the reason why we went with the approach in:
https://mail.python.org/pipermail/python-ideas/2012-May/014969.html
so i'm afraid only option #2 could be somehow viable, but that means the application code has to do the trick you have highlighted.
Unfortunately i don't have any better though to put on the plate.

**Comment by JonasKs** [comment]
@giovannimeo , thank you for the replies! :blush: 

Creating objects in ACI using the SDK is _slick_, as it greatly reduces the amount of JSON I'm used to when using RAW APIs.

I see in the `modelimpl` that there are validators to the fields (such as max length, regex patterns etc), but they don't seem to actually do anything, which is a shame. I have to commit the changes and then see ACI fail.

I'll have to see if I can find a good middle ground of this library in something I can create my self, since I'm personally only interested in ~1-2% of all the objects. Also async support would be nice.

As for generation of the `modelimpl`, do you know if `insieme` is used for that on the APIC? Do you know if there is any API documentation for that library for third parties such as me?

**Comment by giovannimeo** [comment]
@JonasKs, 

let me reply inline about your concerns .... 

> @giovannimeo , thank you for the replies! blush
> 
> Creating objects in ACI using the SDK is _slick_, as it greatly reduces the amount of JSON I'm used to when using RAW APIs.

I totally understand your point, which is the reason why an SDK is useful.

> I see in the `modelimpl` that there are validators to the fields (such as max length, regex patterns etc), but they don't seem to actually do anything, which is a shame. I have to commit the changes and then see ACI fail.

This is something never saw the light unfortunately, we created the provision but we never did implement it, and i will explain in a while.

> I'll have to see if I can find a good middle ground of this library in something I can create my self, since I'm personally only interested in ~1-2% of all the objects. Also async support would be nice.

> As for generation of the `modelimpl`, do you know if `insieme` is used for that on the APIC? 

Yes it's used mainly for the NXOS like CLI project, if you login as admin on APIC you will see the familiar code under:

/controller/yaci/config/
/controller/yaci/execmode/

that implements the NXOS like CLI behavior and uses solely the cobra SDK. I created few of those CLIs, but with the time we realized few pitfalls of the cobra SDK:

1) Every time you upgrade the APIC you have to upgrade the SDK, even though most of the functionality actually stay the same and only the model has been updated.
2) The model classes are very big python files because they threat the model as code, causing issues like the ones we have seen and that forced us to have a modelimpl vs a model directory in acimodel.

Instead @kprav33n came up with a different paradigm, much more resilient and robust, the pyaci paradigm, which is implemented into:

https://github.com/datacenter/pyaci

at the heart of the paradigm shift is the fact that the model is not threaded as code, but is threaded as data. The fact that the model is seen as data implies that you can have exactly the same SDK that can be used against different versions as long as you download the latest and greatest metadata that represent the model. The model representation is summarized in a file called aci-meta.json which was the initial representation of the model, but on that we are actually extending a lot the info we are adding for the use from third-party. The aci-meta.json starting from 5.1 has been directly embedded on the controller at the address:

https://\<apic\>/doc/jsonmeta/aci-meta.json

but is available also from Cisco CCO from here:

https://pubhub.devnetcloud.com/media/model-doc-521/docs/doc/jsonmeta/aci-meta.json

our goal is to have the aci-meta.json as an index that can be used for machine to machine communication. Essentially a program can read these metadata and adapt at the capability of the controller not simply looking at the version. In fact there you have metadata describing what properties or MO are supported, what is being deprecated and what will no longer be accepted. 

> Do you know if there is any API documentation for that library for third parties such as me?

Starting from 5.1 we have revamped the APIC documentation, you can get a view of it from here:

https://developer.cisco.com/site/apic-mim-ref-api/?version=5.2(1)

what has changed is not simply the graphic, but what is behind, the graphic is just a renderer but what is the difference is in fact that we are exporting the documentation in JSON format such that documentation can be used in automatic way by whoever wants to integrate with APIC, and the format is the one defined by aci-meta,json as base, just specialized of a per-class version. Let me explain better, at:

https://pubhub.devnetcloud.com/media/model-doc-521/docs/doc/jsonmeta/aci-meta.json

you have the JSON that you can use directly with PYACI, but if you want to get extra information about a given MO for example fvTenant that is available at:

https://pubhub.devnetcloud.com/media/model-doc-521/docs/doc/jsonmeta/fv/Tenant.json

in general at:

https://pubhub.devnetcloud.com/media/model-doc-521/docs/doc/jsonmeta/\<namespace\>/\<MO\>.json

but the nice thing is that you can have it also directy at the address:

https://\<apic\>/doc/jsonmeta/\<namespace\>/\<MO\>.json

so you can probe dynamically what an APIC supports.

Bottom line we keep shipping Cobra because customers like you use it, but we are trying to invest on the pyaci direction more because allow more flexible description for machine to machine interaction. Also PYACI is only available on github, we don't have an internal repository so the source code is fully open source. The model is instead published directly from the controller or via devnet.

Now that said, even with this new paradigm, the auto-completion for python is not available, because the model will be seen as data instead of code. Can this metadata be used for python autocompletion, maybe yes, but in all honesty i never thought to implement one, i usually scrub through the json files for what i need.
 
Regarding the asyncio support, this is also missing in PYACI, because of lack of time, but also because asyncio for ACI way of configuration is less compelling. With ACI you don't necessarely have to POST a single MO at the time, but usually you form a configuration that span multiple lines, in fact that is the best usage of ACI, because when you do so, the configuration is threaded as a single transaction and is validated for inconsistency ...  if is inconsistent the whole is not applied.

Sorry for the long explanation, but i thought it would be nice to give an overview of where we are at.

**Comment by JonasKs** [comment]
@giovannimeo, thank you so much for taking the time to explain everything!

> Yes it's used mainly for the NXOS like CLI project, if you login as admin on APIC you will see the familiar code under:
> ...

Thank you! I'll check it out.

> Instead @kprav33n came up with a different paradigm, much more resilient and robust, the pyaci paradigm, which is implemented into:
> ...

This project actually was how I found the `insieme` module on the APIC. I've generated `pydantic` models based on the meta-file created through that project (with some alterations). Thanks for describing the structure of the meta files.

 
> Bottom line we keep shipping Cobra because customers like you use it, but we are trying to invest on the pyaci direction more because allow more flexible description for machine to machine interaction.

I don't use it my self as of today (I use raw JSON), but I see that the [`DCAUTO 300-635`](https://developer.cisco.com/certification/exam-topic-dc-auto/) exam which I intend to take after `DEVCOR`, actually requires to use the Cobra SDK. I came here mostly because I wasn't blessed after giving it its second chance. :grin: 

I'm excited to see that you want to invest on the `pyaci` direction. I'll make sure to give it a go ASAP. 

> Regarding the asyncio support, this is also missing in PYACI, because of lack of time, but also because asyncio for ACI way of configuration is less compelling. 

I agree on this, and I do construct huge JSONs which I then post. How ever, since we have a micro service architecture using FastAPI, all fetch operations are done async. We dynamically chose which values to configure based on tagged objects in ACI, what exist in the common tenant etc. This often ends up to be quite a lot of API calls, even though ACI is quick, it is nice to do them async, and we can't really mix the sync world into it.

Me and my collegue actually did a "DevNet create"-talk on how we visualize tenant information in ACI, using pydantic and FastAPI. If you're interested you can check it out [here](https://www.youtube.com/watch?v=1gjMay47DdU&t). Demo starts around 1:15.

>  Sorry for the long explanation, but i thought it would be nice to give an overview of where we are at.

Thank you, really!

------------------

Since we're at it, I'd like to ask one more question, but I understand if you can't answer it: We're struggling to reliably use WebSockets to monitor events in ACI.

If we wanted to know about all created Application Profiles, we could subscribe to this data in two ways:

#### 1) Using the normal `fvAP` class, with subscriptions:  

URL: `https://apic-ip-address/api/class/fvAP.json?subscription=yes`


This query returns data such as:
```py
{
  'subscriptionId': [
    '72059964880715777'
  ],
  'imdata': [
    {
      'fvAp': {
        'attributes': {
          'annotation': '',
          'childAction': '',
          'descr': '',
          'dn': 'uni/tn-JonasPres/ap-Test-AP',
          'extMngdBy': '',
          'lcOwn': 'local',
          'modTs': '2021-06-22T15:47:16.728+02:00',
          'monPolDn': 'uni/tn-common/monepg-default',
          'name': 'Test-AP',
          'nameAlias': '',
          'ownerKey': '',
          'ownerTag': '',
          'prio': 'unspecified',
          'rn': '',
          'status': 'created',
          'uid': '15464'
        }
      }
    }
  ]
}
```


This works perfectly fine, but we have no control over potentially lost events. This is because there is no id connected to the event.

In other words, if our application restarts or one of the APICs are down for maintenance, we could lose events, and we have no way of re-fetching that event.

#### 2) The solution today is to use the audit log, subscribing to the aaaModLR class:

URL: `https://apic-ip-address/api/node/class/aaaModLR.json?subscription=yes&query-target-filter=eq(aaaModLR.code,"E4211942")`  
(The code E4211942 here means that an AP is created.)  


This query returns events such as:

```py
{
  'subscriptionId': [
    '72059964880715785'
  ],
  'imdata': [
    {
      'aaaModLR': {
        'attributes': {
          'affected': 'uni/tn-JonasPres/ap-Test-AP',
          'cause': 'transition',
          'changeSet': 'name:Test-AP, prio:unspecified',
          'childAction': '',
          'clientTag': '',
          'code': 'E4211942',
          'created': '2021-06-22T15:47:18.211+02:00',
          'descr': 'Ap Test-AP created',
          'dn': 'subj-[uni/tn-JonasPres/ap-Test-AP]/mod-4295026262',
 ------>  'id': '4295026262',  <----------------------------------------------
          'ind': 'creation',
          'rn': '',
          'sessionId': 'KZd/aPhpSR6H2mm3Z/7yYg==',
          'severity': 'info',
          'status': 'created',
          'trig': 'config',
          'txId': '576460752335996871',
          'user': 'aa545'
        }
      }
    }
  ]
}

```


This response does have an id, `4295026262`. This means, if our application crashes, all we need to know is the latest ID we received.



When our connection is established to the APIC again, we can query for events newer than this event: `https://apic-ip-address/api/node/class/aaaModLR.json?query-target-filter=and(eq(aaaModLR.code,"E4211942")gt(aaaModLR.id,"4295026261"))&order-by=aaaModLR.created|desc`

We now have a way to ensure we never lose messages from ACI.



The only downside with this, is that the `changeSet` isn't structured as the data we get from the `fvAp` subscription.

It is possible to parse out with regex, but very annoying since I must account for descriptions(and other objects) being in the same format as the `changeSet` string.

Looking at the response above, we can see the changeSet:

> '**descr:contact name:Jonas, if not available: Torbjorn**, name:Test2-AP, prio:unspecified'



Here, the actual description is `contact name:Jonas, if not available: Torbjorn`, but it contains the word `name:`. This is the same standard `changedSet` has for the name of the AP that was created: `name:Test2-AP,`. This kind of behaviour is hard to work around and will quickly cause bugs.


I have a feature request to attach IDs for events when subscribing to e.g. `class/fvAp` or to get structured data from the `aaaModLr` events (not sure how that is doing though), but I was wondering if you have any other smarter solution I could check out?

**Comment by giovannimeo** [comment]
> @giovannimeo, thank you so much for taking the time to explain everything!
> 
> > Yes it's used mainly for the NXOS like CLI project, if you login as admin on APIC you will see the familiar code under:
> > ...
> 
> Thank you! I'll check it out.
> 
> > Instead @kprav33n came up with a different paradigm, much more resilient and robust, the pyaci paradigm, which is implemented into:
> > ...
> 
> This project actually was how I found the `insieme` module on the APIC. I've generated `pydantic` models based on the meta-file created through that project (with some alterations). Thanks for describing the structure of the meta files.
> 
> > Bottom line we keep shipping Cobra because customers like you use it, but we are trying to invest on the pyaci direction more because allow more flexible description for machine to machine interaction.
> 
> I don't use it my self as of today (I use raw JSON), but I see that the [`DCAUTO 300-635`](https://developer.cisco.com/certification/exam-topic-dc-auto/) exam which I intend to take after `DEVCOR`, actually requires to use the Cobra SDK. I came here mostly because I wasn't blessed after giving it its second chance. grin

Oh i didn't know we had an exam on cobra ... will try to see how to adjust that direction.
 
> I'm excited to see that you want to invest on the `pyaci` direction. I'll make sure to give it a go ASAP.
> 
> > Regarding the asyncio support, this is also missing in PYACI, because of lack of time, but also because asyncio for ACI way of configuration is less compelling.
> 
> I agree on this, and I do construct huge JSONs which I then post. How ever, since we have a micro service architecture using FastAPI, all fetch operations are done async. We dynamically chose which values to configure based on tagged objects in ACI, what exist in the common tenant etc. This often ends up to be quite a lot of API calls, even though ACI is quick, it is nice to do them async, and we can't really mix the sync world into it.

I cannot promise we can do anything for pyaci as well, but if you guys want to contribute something, happy to evaluate it.
 
> Me and my collegue actually did a "DevNet create"-talk on how we visualize tenant information in ACI, using pydantic and FastAPI. If you're interested you can check it out [here](https://www.youtube.com/watch?v=1gjMay47DdU&t). Demo starts around 1:15.

The work you guys have done is quite amazing i must say. I'm sure the jsonmeta we are exporting will be quite useful for you guys to create your pydantic model.

> > Sorry for the long explanation, but i thought it would be nice to give an overview of where we are at.
> 
> Thank you, really!
> 
> Since we're at it, I'd like to ask one more question, but I understand if you can't answer it: We're struggling to reliably use WebSockets to monitor events in ACI.
> 
> If we wanted to know about all created Application Profiles, we could subscribe to this data in two ways:
> #### 1) Using the normal `fvAP` class, with subscriptions:
> 
> URL: `https://apic-ip-address/api/class/fvAP.json?subscription=yes`
> 
> This query returns data such as:
> 
> ```python
> {
>   'subscriptionId': [
>     '72059964880715777'
>   ],
>   'imdata': [
>     {
>       'fvAp': {
>         'attributes': {
>           'annotation': '',
>           'childAction': '',
>           'descr': '',
>           'dn': 'uni/tn-JonasPres/ap-Test-AP',
>           'extMngdBy': '',
>           'lcOwn': 'local',
>           'modTs': '2021-06-22T15:47:16.728+02:00',
>           'monPolDn': 'uni/tn-common/monepg-default',
>           'name': 'Test-AP',
>           'nameAlias': '',
>           'ownerKey': '',
>           'ownerTag': '',
>           'prio': 'unspecified',
>           'rn': '',
>           'status': 'created',
>           'uid': '15464'
>         }
>       }
>     }
>   ]
> }
> ```
> 
> This works perfectly fine, but we have no control over potentially lost events. This is because there is no id connected to the event.
> 
> In other words, if our application restarts or one of the APICs are down for maintenance, we could lose events, and we have no way of re-fetching that event.

> #### 2) The solution today is to use the audit log, subscribing to the aaaModLR class:
> 
> URL: `https://apic-ip-address/api/node/class/aaaModLR.json?subscription=yes&query-target-filter=eq(aaaModLR.code,"E4211942")` (The code E4211942 here means that an AP is created.)
> 
> This query returns events such as:
> 
> ```python
> {
>   'subscriptionId': [
>     '72059964880715785'
>   ],
>   'imdata': [
>     {
>       'aaaModLR': {
>         'attributes': {
>           'affected': 'uni/tn-JonasPres/ap-Test-AP',
>           'cause': 'transition',
>           'changeSet': 'name:Test-AP, prio:unspecified',
>           'childAction': '',
>           'clientTag': '',
>           'code': 'E4211942',
>           'created': '2021-06-22T15:47:18.211+02:00',
>           'descr': 'Ap Test-AP created',
>           'dn': 'subj-[uni/tn-JonasPres/ap-Test-AP]/mod-4295026262',
>  ------>  'id': '4295026262',  <----------------------------------------------
>           'ind': 'creation',
>           'rn': '',
>           'sessionId': 'KZd/aPhpSR6H2mm3Z/7yYg==',
>           'severity': 'info',
>           'status': 'created',
>           'trig': 'config',
>           'txId': '576460752335996871',
>           'user': 'aa545'
>         }
>       }
>     }
>   ]
> }
> ```
> 
> This response does have an id, `4295026262`. This means, if our application crashes, all we need to know is the latest ID we received.
> 
> When our connection is established to the APIC again, we can query for events newer than this event: `https://apic-ip-address/api/node/class/aaaModLR.json?query-target-filter=and(eq(aaaModLR.code,"E4211942")gt(aaaModLR.id,"4295026261"))&order-by=aaaModLR.created|desc`
> 
> We now have a way to ensure we never lose messages from ACI.
> 
> The only downside with this, is that the `changeSet` isn't structured as the data we get from the `fvAp` subscription.
> 
> It is possible to parse out with regex, but very annoying since I must account for descriptions(and other objects) being in the same format as the `changeSet` string.
> 
> Looking at the response above, we can see the changeSet:
> 
> > '**descr:contact name:Jonas, if not available: Torbjorn**, name:Test2-AP, prio:unspecified'
> 
> Here, the actual description is `contact name:Jonas, if not available: Torbjorn`, but it contains the word `name:`. This is the same standard `changedSet` has for the name of the AP that was created: `name:Test2-AP,`. This kind of behaviour is hard to work around and will quickly cause bugs.
> 
> I have a feature request to attach IDs for events when subscribing to e.g. `class/fvAp` or to get structured data from the `aaaModLr` events (not sure how that is doing though), but I was wondering if you have any other smarter solution I could check out?

Have you looked at the approach taken by the ACI Enhanced Endpoint Tracker:

https://github.com/agccie/ACI-EnhancedEndpointTracker/blob/master/Service/app/models/aci/subscription_ctrl.py

where they monitor the liveness of the subscription and when they realize it was stale because the app restarted or the controller restarted, they do a fresh bulk re-read. It works pretty well, i have been using Enhanced Endpoint Tracker for close to 3 years and it has served me well. The mechanism there is pretty robust from my experience, and also logically seems pretty sound.

**Comment by JonasKs** [comment]
> I cannot promise we can do anything for pyaci as well, but if you guys want to contribute something, happy to evaluate it.

I'll keep that in mind, thank you!

> The work you guys have done is quite amazing i must say. I'm sure the jsonmeta we are exporting will be quite useful for you guys to create your pydantic model.

Thanks 😊 

> Have you looked at the approach taken by the ACI Enhanced Endpoint Tracker:
> ...

No, I have not, but I will check that out and see if we can take some inspiration. Looks neat!


Have a great week, thanks again for taking the time. ☺️
